### DISCUSSION TITLE
guides/knowledge-distillation/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#24946] [Docs]
# guides/knowledge-distillation/

Master knowledge distillation for Ultralytics YOLO to optimize student model performance with a teacher model.

https://docs.ultralytics.com/guides/knowledge-distillation

**Comment by UltralyticsAssistant** [comment]
👋 Hello, thank you for your interest in Ultralytics 🚀! This is an automated response to help get things started, and an Ultralytics engineer will also assist soon. Since this discussion is about documentation for Ultralytics YOLO, we recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples, and many common questions may already be answered.
If this is a 🐛 bug report related to the guide or its examples, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) to help us debug it.
If this is a custom training or knowledge distillation ❓ question, please provide as much information as possible, including dataset image examples, training logs, your teacher/student model setup, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.
## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
## Environments
YOLO may be run in any of the following up-to-date verified environments, with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by andyrey** [comment]
Very interesting and useful, thank you!

**Reply by glenn-jocher** [reply]
Thanks @andyrey, glad it’s useful! If you or others have questions or examples from trying `distill_model`, please drop them here so the Ultralytics team and community can see and discuss them.

**Comment by CharbelMsallem** [comment]
Cool! If same dataset, can the class mapping or the number of classes differ from parent to student but using the same images and split?

**Reply by CharbelMsallem** [reply]
I mean classes of student be a subset of parent classes

**Reply by glenn-jocher** [reply]
Same images/splits can be reused, but for the supported distillation path the teacher and student should use the same task and `data.yaml` class configuration (`nc`, `names`, and class indices), per the [Knowledge Distillation guide](https://docs.ultralytics.com/guides/knowledge-distillation). If the student should learn only a subset, create a subset dataset/YAML with labels remapped to contiguous IDs and train or fine-tune a teacher on that same subset/mapping before using it as `distill_model`.

**Reply by CharbelMsallem** [reply]
okat thanks!

**Reply by glenn-jocher** [reply]
You're welcome! Please feel free to drop any follow-up questions in this discussion so the Ultralytics team and community can see them.
